rustworkx-core/src/token_swapper.rs

```diff
@@ -0,0 +1,556 @@
+        let num_nodes = self.graph.node_count();
+        let num_edges = self.graph.edge_count();
+
+        // Directed graph with nodes matching ``graph`` and
+        // edges for neighbors closer than nodes
+        let mut digraph = StableGraph::with_capacity(num_nodes, num_edges);
+        // Same as digraph with no self edges
+        let mut sub_digraph = StableGraph::with_capacity(num_nodes, num_edges);
+        // A list of nodes that are remaining to be tried
+        let mut todo_nodes = Vec::with_capacity(self.input_mapping.len());
+
+        let mut rng_seed: Pcg64 = match self.rng_seed {
+            Some(rng_seed) => Pcg64::seed_from_u64(rng_seed),
+            None => Pcg64::from_entropy(),
+        };
+        // Add nodes to the digraph/sub_digraph and build the node maps
+        for node in self.graph.node_identifiers() {
+            let index = digraph.add_node(());
+            sub_digraph.add_node(());
+            self.node_map.insert(node, index);
+            self.rev_node_map.insert(index, node);
+        }
+        // The input mapping in HashMap form using NodeIndex
+        let mut tokens: HashMap<NodeIndex, NodeIndex> = self
+            .input_mapping
+            .iter()
+            .map(|(k, v)| (self.node_map[k], self.node_map[v]))
+            .collect();
+        // todo_nodes are all the mapping entries where left != right
+        for (node, dest) in &tokens {
+            if node != dest {
+                todo_nodes.push(*node);
+            }
+        }
+        // Add initial edges to the digraph/sub_digraph
+        for node in self.graph.node_identifiers() {
+            self.add_token_edges(
+                self.node_map[&node],
+                &mut digraph,
+                &mut sub_digraph,
+                &mut tokens,
+            );
+        }
+        // Do the trials and build a results Vec
+        let mut trial_results: Vec<Vec<Swap>> = Vec::new();
+        for _ in 0..self.trials {
+            let results = self.trial_map(
+                &mut digraph.clone(),
+                &mut sub_digraph.clone(),
+                &mut tokens.clone(),
+                &mut todo_nodes.clone(),
+                &mut rng_seed,
+            );
+            trial_results.push(results);
+        }
+        // Build the first results Vec until a 0 len result is found
+        let mut first_results: Vec<Vec<Swap>> = Vec::new();
+        for result in trial_results {
+            if result.is_empty() {
+                first_results.push(result);
+                break;
+            }
+            first_results.push(result);
+        }
+        // Return the min of the first results
+        let mut res_min = MAX;
+        let mut final_result: Vec<Swap> = Vec::new();
+        for res in first_results {
+            let res_len = res.len();
+            if res_len < res_min {
+                final_result = res;
+                res_min = res_len;
+            }
+        }
+        final_result
```

Comment:
Heh, this can also be combined into the same iterator chain above. You can use `.min_by_key()` instead of collect and avoid all the Vec allocations for results. This makes the full iterator chain something like:

```rust
     let mut empty_found = false;
     let final_result Vec<Swap> = (0..self.trials)
            .map(|_| {
                let results = self.trial_map(
                    &mut digraph.clone(),
                    &mut sub_digraph.clone(),
                    &mut tokens.clone(),
                    &mut todo_nodes.clone(),
                   &mut rng_seed,
                )
        })
        .filter(|result| {
            if empty_found {
                None
            } else {
               if x.is_empty() {
                    empty_found = true;
               }
                Some(x)
            }
        })
        .min_by_key(|result| result.len())
        .unwrap()
```

Reply:
Still working on this. Will need a little refinement.

Reply:
The closure for the `filter` method has to return a bool. In the previous code, I believe the only reason for checking for `is_empty` is to break out of the `for` early since an empty will be the min. Breaking out of an iterator is not so easy as best I can tell.

This code works and I think it's pretty efficient without the `is_empty` unless `trials` is a very large number,
```
        (0..self.trials)
            .map(|_| {
                self.trial_map(
                    digraph.clone(),
                    sub_digraph.clone(),
                    tokens.clone(),
                    todo_nodes.clone(),
                    &mut rng_seed,
                )
            })
            .min_by_key(|result| result.len())
            .unwrap()
```

Reply:
Sure, that makes sense especially for small numbers of `trials`. 

The other thing which I think might be worth looking into here is parallelism. Since this iterator is quite compact and has no shared data dependency (because of the `clone()` calls we can easily use rayon here and potentially further speed up this iterator. It should be mostly a matter of adding a `.into_par_iter()` before the map (although we'll have to adjust the seed usage so it's not a shared value across all the threads.

Reply:
Ok. Will look into the parallel.

Reply:
Ok. Got it working. Now with my local test,
- Old token_swapper: 16.7 sec
- New without parallel: 7.6 sec
- New with parallel: 2.6 sec

I'll update and push shortly.

Reply:
Done in https://github.com/Qiskit/rustworkx/pull/765/commits/83ccd8c397727df14e27164e3ede77d84243086b.